File: lib/controllers/admin_controller.dart
import 'package:get/get.dart';
import 'package:http/http.dart' as http;
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class AdminController extends GetxController {
  var isLoading = false.obs;
  var totalUsers = 0.obs;
  var totalOrders = 0.obs;
  var activeOrders = 0.obs;
  var totalRevenue = 0.0.obs;
  var recentOrders = [].obs;

  var users = [].obs;
  var isUsersLoading = false.obs;

  var orders = [].obs;
  var isOrdersLoading = false.obs;

  var reviews = [].obs;
  var isReviewsLoading = false.obs;

  @override
  void onInit() {
    super.onInit();
    fetchDashboardData();
  }

  Future<String?> _getToken() async {
    final prefs = await SharedPreferences.getInstance();
    return prefs.getString('token');
  }

  Map<String, String> _headers(String token) => {
        'Authorization': 'Bearer $token',
        'Accept': 'application/json',
        'Content-Type': 'application/json',
      };

  // ========== DASHBOARD ==========
  Future<void> fetchDashboardData() async {
    isLoading.value = true;
    try {
      final token = await _getToken();
      if (token == null) { isLoading.value = false; return; }

      final response = await http.get(
        Uri.parse('http://127.0.0.1:8000/api/admin/dashboard'),
        headers: _headers(token),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        totalUsers.value = data['total_users'] ?? 0;
        totalOrders.value = data['total_orders'] ?? 0;
        activeOrders.value = data['active_orders'] ?? 0;
        totalRevenue.value = double.tryParse(data['total_revenue'].toString()) ?? 0.0;
        recentOrders.value = data['recent_orders'] ?? [];
      } else {
        Get.snackbar('Error', 'Server: ${response.statusCode}');
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal: $e');
    } finally {
      isLoading.value = false;
    }
  }

  // ========== USERS ==========
  Future<void> fetchUsers() async {
    isUsersLoading.value = true;
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.get(
        Uri.parse('http://127.0.0.1:8000/api/admin/users'),
        headers: _headers(token),
      );
      if (response.statusCode == 200) {
        users.value = jsonDecode(response.body);
      }
    } catch (e) {
      Get.snackbar('Error', 'Gagal memuat pengguna');
    } finally {
      isUsersLoading.value = false;
    }
  }

  Future<void> updateUserRole(int userId, String newRole) async {
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.put(
        Uri.parse('http://127.0.0.1:8000/api/admin/users/$userId'),
        headers: _headers(token),
        body: jsonEncode({'role': newRole}),
      );
      if (response.statusCode == 200) {
        Get.snackbar('Berhasil', 'Role pengguna diperbarui');
        fetchUsers();
      } else {
        Get.snackbar('Error', 'Gagal memperbarui role');
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal');
    }
  }

  Future<void> deleteUser(int userId) async {
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.delete(
        Uri.parse('http://127.0.0.1:8000/api/admin/users/$userId'),
        headers: _headers(token),
      );
      if (response.statusCode == 200) {
        Get.snackbar('Berhasil', 'Pengguna dihapus');
        fetchUsers();
        fetchDashboardData();
      } else {
        Get.snackbar('Error', 'Gagal menghapus pengguna');
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal');
    }
  }

  // ========== ORDERS ==========
  Future<void> fetchOrders() async {
    isOrdersLoading.value = true;
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.get(
        Uri.parse('http://127.0.0.1:8000/api/admin/orders'),
        headers: _headers(token),
      );
      if (response.statusCode == 200) {
        orders.value = jsonDecode(response.body);
      }
    } catch (e) {
      Get.snackbar('Error', 'Gagal memuat pesanan');
    } finally {
      isOrdersLoading.value = false;
    }
  }

  Future<void> updateOrderStatus(int orderId, String status) async {
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.put(
        Uri.parse('http://127.0.0.1:8000/api/admin/orders/$orderId/status'),
        headers: _headers(token),
        body: jsonEncode({'status': status}),
      );
      if (response.statusCode == 200) {
        Get.snackbar('Berhasil', 'Status pesanan diperbarui');
        fetchOrders();
        fetchDashboardData();
      } else {
        Get.snackbar('Error', 'Gagal memperbarui status');
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal');
    }
  }

  Future<void> deleteOrder(int orderId) async {
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.delete(
        Uri.parse('http://127.0.0.1:8000/api/admin/orders/$orderId'),
        headers: _headers(token),
      );
      if (response.statusCode == 200) {
        Get.snackbar('Berhasil', 'Pesanan dihapus');
        fetchOrders();
        fetchDashboardData();
      } else {
        Get.snackbar('Error', 'Gagal menghapus pesanan');
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal');
    }
  }

  // ========== REVIEWS ==========
  Future<void> fetchReviews() async {
    isReviewsLoading.value = true;
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.get(
        Uri.parse('http://127.0.0.1:8000/api/admin/reviews'),
        headers: _headers(token),
      );
      if (response.statusCode == 200) {
        reviews.value = jsonDecode(response.body);
      }
    } catch (e) {
      Get.snackbar('Error', 'Gagal memuat ulasan');
    } finally {
      isReviewsLoading.value = false;
    }
  }

  Future<void> deleteReview(int reviewId) async {
    try {
      final token = await _getToken();
      if (token == null) return;
      final response = await http.delete(
        Uri.parse('http://127.0.0.1:8000/api/admin/reviews/$reviewId'),
        headers: _headers(token),
      );
      if (response.statusCode == 200) {
        Get.snackbar('Berhasil', 'Ulasan dihapus');
        fetchReviews();
      } else {
        Get.snackbar('Error', 'Gagal menghapus ulasan');
      }
    } catch (e) {
      Get.snackbar('Error', 'Koneksi gagal');
    }
  }
}

File: nearty-backend/app/Http/Controllers/api/AdminController.php
<?php

namespace App\Http\Controllers\Api;

use App\Http\Controllers\Controller;
use App\Models\Order;
use App\Models\Review;
use App\Models\User;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Validator;

class AdminController extends Controller
{
    // ========== DASHBOARD ==========
    // GET /api/admin/dashboard
    public function dashboard()
    {
        return response()->json([
            'total_users' => User::count(),
            'total_orders' => Order::count(),
            'active_orders' => Order::whereIn('status', ['pending', 'accepted', 'picked_up'])->count(),
            'total_revenue' => Order::where('status', 'completed')->sum('fee'),
            'recent_orders' => Order::with('user')->latest()->take(5)->get(),
        ]);
    }

    // ========== USERS ==========
    // GET /api/admin/users
    public function users()
    {
        return response()->json(User::latest()->get());
    }

    // PUT /api/admin/users/{id}
    public function updateUserRole(Request $request, User $user)
    {
        $validator = Validator::make($request->all(), [
            'role' => ['required', 'string', 'in:user,driver,admin'],
        ]);

        if ($validator->fails()) {
            return response()->json(['message' => 'Gagal memperbarui role'], 422);
        }

        $user->update(['role' => $request->role]);

        return response()->json($user);
    }

    // DELETE /api/admin/users/{id}
    public function deleteUser(User $user)
    {
        $user->delete();

        return response()->json(['message' => 'Pengguna dihapus']);
    }

    // ========== ORDERS ==========
    // GET /api/admin/orders
    public function orders()
    {
        return response()->json(
            Order::with(['user', 'driver', 'review'])->latest()->get()
        );
    }

    // PUT /api/admin/orders/{id}/status
    public function updateOrderStatus(Request $request, Order $order)
    {
        $validator = Validator::make($request->all(), [
            'status' => ['required', 'string', 'in:pending,accepted,picked_up,completed,cancelled'],
        ]);

        if ($validator->fails()) {
            return response()->json(['message' => 'Gagal memperbarui status'], 422);
        }

        $order->update(['status' => $request->status]);

        return response()->json($order->fresh(['user', 'driver']));
    }

    // DELETE /api/admin/orders/{id}
    public function deleteOrder(Order $order)
    {
        $order->delete();

        return response()->json(['message' => 'Pesanan dihapus']);
    }

    // ========== REVIEWS ==========
    // GET /api/admin/reviews
    public function reviews()
    {
        return response()->json(
            Review::with('order.user')->latest()->get()
        );
    }

    // DELETE /api/admin/reviews/{id}
    public function deleteReview(Review $review)
    {
        $review->delete();

        return response()->json(['message' => 'Ulasan dihapus']);
    }
}
